compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -0,0 +1,242 @@
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(begin));
+        template_vec.extend(template.iter().cloned());
+        template_vec.push(rustc_ast::ast::InlineAsmTemplatePiece::String(end));
+
+        cx.codegen_global_asm(&template_vec, &operands, options, line_spans);
+    }
+
+    fn inline_to_global_operand(&self, op: &InlineAsmOperand<'tcx>) -> GlobalAsmOperandRef<'tcx> {
+        match op {
+            InlineAsmOperand::Const { value } => {
+                let const_value = self.eval_mir_constant(value);
+                let string = common::asm_const_to_str(
+                    self.cx.tcx(),
+                    value.span,
+                    const_value,
+                    self.cx.layout_of(value.ty()),
+                );
+                GlobalAsmOperandRef::Const { string }
+            }
+            InlineAsmOperand::SymFn { value } => {
+                let instance = match value.ty().kind() {
+                    &ty::FnDef(def_id, args) => Instance::new(def_id, args),
+                    _ => bug!("asm sym is not a function"),
+                };
+
+                GlobalAsmOperandRef::SymFn { instance }
+            }
+            InlineAsmOperand::SymStatic { def_id } => {
+                GlobalAsmOperandRef::SymStatic { def_id: *def_id }
+            }
+            InlineAsmOperand::In { .. }
+            | InlineAsmOperand::Out { .. }
+            | InlineAsmOperand::InOut { .. }
+            | InlineAsmOperand::Label { .. } => {
+                bug!("invalid operand type for naked_asm!")
+            }
+        }
+    }
+}
+
+enum AsmBinaryFormat {
+    Elf,
+    Macho,
+    Coff,
+}
+
+impl AsmBinaryFormat {
+    fn from_target(target: &rustc_target::spec::Target) -> Self {
+        if target.is_like_windows {
+            Self::Coff
+        } else if target.options.vendor == "apple" {
+            Self::Macho
+        } else {
+            Self::Elf
+        }
+    }
+}
+
+fn linkage_directive(linkage: Linkage) -> Option<&'static str> {
+    if true {
+        // this is required. emitting nothing or .weak will emit
+        //
+        // > Global is external, but doesn't have external or weak linkage!
+        //
+        // and then aborts compilation
+        return Some(".globl");
+    }
+
+    match linkage {
+        Linkage::External => Some(".globl"),
+        Linkage::WeakAny | Linkage::WeakODR => Some(".weak"),
+        Linkage::LinkOnceAny | Linkage::LinkOnceODR => Some(".weak"),
+        Linkage::Internal | Linkage::Private => None, // just doesn't emit any attribute
+        Linkage::ExternalWeak => None,                // terminates with sigill on godbolt
+        Linkage::AvailableExternally => None,         // does not even emit the definition
+        Linkage::Appending => None,                   // only valid on global variables
+        Linkage::Common => None,                      // function may not have common linkage
```

Comment:
These should probably just panic as they should be unreachable.

Reply:
but is this function even meaningful now that we effectively force `Linkage::External` with `InstantiationMode::GloballyShared`? A user can manually override with`#[linkage = "weak"]`, but that will just crate llvm errors like this

```rust
#[naked]
#[linkage = "weak"]
extern "C" fn generic_naked<const N: u64>() -> u64 {
    unsafe {
        asm!(
            "mov rax, {}",
            "ret",
            const N,
            options(noreturn)
        );
    }
}
```

```
Global is external, but doesn't have external or weak linkage!
ptr @_ZN14naked_function13generic_naked17h0027c5bd3c9adc32E
Global is external, but doesn't have external or weak linkage!
ptr @_ZN14naked_function13generic_naked17h97e15019324bc8d7E
rustc-LLVM ERROR: Broken module found, compilation aborted!
```

Reply:
actually my other PR https://github.com/rust-lang/rust/pull/127853 considers `#[naked]` incompatible with `#[linkage]`, so this function can just go.

Reply:
also the llvm errors only occur when running `cargo +stage1 rustc -- --emit=llvm-ir`, `cargo run` works just fine.

Reply:
external linkage and emitting `.globl` will now always happen. I believe it is the only way to make linking the extern function with the global asm block work reliably.